### DISCUSSION TITLE
NMEA messages reading via PySerial with Ntrip ?

### DISCUSSION THREAD

**Discussion by jehlijos** [semuconsulting/PyGPSClient#97] [Q&A]
Hi,
I am kind of new to GNSS technology. I am reading NMEA messages from ublox with python script on RaspberryPi (via PySerial). I need to find a way to connect this ublox to ntrip and also read data from USB port every second.
(https://github.com/jehlijos/GNSS-EGNOS-EDAS-public_copy)

Is it possible to set PyGPSClient to send ntrip to ublox while usb ports are used to send nmea messages to Raspberry (via python). Or are there any alternatives to make this work?

Thanks for any inputs in advance.

**Comment by semuadmin** [comment]
Hi @jehlijos 

It's not clear what your intention is here, and I'm not in a position to review the code you've linked to above (which I note uses the `pynmea2` library rather than the `pynmeagps` library used by PyGPSClient).

Can you explain more clearly what you're ultimately looking to achieve? A diagram would be helpful if you have one.

In answer to your question:
1. Yes, PyGPSClient implements an [NTRIP client](https://github.com/semuconsulting/PyGPSClient#ntripconfig) which can receive NTRIP (RTCM) data from an NTRIP caster and send this to any NTRIP-compatible receiver e.g. u-blox ZED-F9P, while simultaneously parsing NMEA messages from that receiver.
1. You could also use the [`gnssntripclient`](https://github.com/semuconsulting/pygnssutils#gnssntripclient) CLI utility (or its underlying Python classes) to achieve the same end without the need for a GUI, if that suits your needs better.

**Reply by jehlijos** [reply]
![schema](https://github.com/semuconsulting/PyGPSClient/assets/117917253/aa659ac0-a338-4857-aa90-712164781c56)
Thanks for reply!
I attached schematic picture of what I am trying to achieve.
But with reciever 3 I am having the problem. Either PyGPSClient or my python script shows same error if other one is running:
"'device reports readiness to read but returned no data ' serial.serialutil.SerialException: device reports readiness to read but returned no data (device disconnected or multiple access on port?)"

Do you think it is possible to make this work with PyGPSClient or do I have to implement the gnssntripclient CLI utility in my own code?

**Comment by semuadmin** [comment]
Hi @jehlijos 

Thanks for the additional information - the diagram makes your intentions clearer.

If I understand correctly, you want to capture NMEA GNSS data from 3 receivers concurrently to compare EGNOS (SBAS) corrected, RTCM (NTRIP) corrected and uncorrected navigation solutions? I'm not sure what that would accomplish - SBAS and NTRIP are fundamentally different GNSS correction protocols and they wouldn't really produce comparable results - but it's perfectly possible to do this as an academic exercise (see, for example, https://www.semuconsulting.com/gnsswiki/ for a concise explanation of how SBAS and NTRIP work).

I'm assuming the 3 receivers are connected on three different RPi serial ports e.g. /dev/ttyACM0, /dev/ttyACM1 and /dev/ttyACM2 (*or something like that*) and that you have confimed you're using the correct baud rates, etc.? I'm also assuming that 'receiver 3' is capable of processing NTRIP corrections e.g. something like the u-blox ZED-F9P.

The error message you're getting is almost certainly because you have more than one process running on your RPi trying to access the same serial port. The usual culprit is something like `gpsd` or another daemon process running in the background - that would be the first thing I would check. Try shutting down all your other scripts and background processes and connecting PyGPSClient to 'receiver 3' on its own.

Provided you can resolve the serial port contention issue, I see no reason why you should not be able to use PyGPSClient's NTRIP client facility, but it's not clear how you intend to get data out of PyGPSClient and into your database? PyGPSClient has a data-logging facility which can record incoming binary GNSS data (NMEA, UBX or RTCM) to a file, but that obviously wouldn't provide you with a real-time data feed.

You may be better adapting the [`rtk_example.py`](https://github.com/semuconsulting/pygnssutils/blob/main/examples/rtk_example.py) example from the pygnssutils library - you could actually adapt this to run 3 concurrent I/O threads - one for each receiver - and output the consolidated data in whatever format you need it in.
